### DISCUSSION TITLE
General debugging tips

### DISCUSSION THREAD

**Discussion by ajensen1234** [CERN/TIGRE#539] [Q&A]
Hi! I'm wondering if there are any general guidelines for debugging that might help to steer toward a good reconstruction! Tangentially, this would be related to the sensitivity of the reconstruction, and how close the geometry needs to be defined to get even a “bad” reconstruction.

For context, I have a collection of cone-beam fluoro images of a hip phantom captured from an OEC-9800:
![hip_phantom](https://github.com/CERN/TIGRE/assets/63751488/71c1e85d-a58b-4d58-b414-8eea96312c80)
![image](https://github.com/CERN/TIGRE/assets/63751488/e443728e-0132-4e1a-9a66-6add75839d2f)
![hip_moving](https://github.com/CERN/TIGRE/assets/63751488/6630bfe9-9a3a-4d0c-908d-53b14f9e24ea)

In order to determine the `geo` parameters, we used motion capture rigidly affixed to both the phantom and the OEC C-ARM to track motion.

Here's a rough view I've generated of the imaging geometry during a “sweep” of the c-arm:
![image](https://github.com/CERN/TIGRE/assets/63751488/3c55e5bc-18c9-4b3e-a152-904a23ed9a51)

And so, back to the original question. How sensitive would the reconstruction (say, CGLS) be of minor perturbations in the geometry definitions?  If I knew the source to detector (DSD) and source to origin (DSO), would just using those be sufficient for generating a “bad” reconstruction? Using all combinations of geometry offsets, center of rotation offsets, etc, the only results I've achieved look like the images attached below:
![cgls](https://github.com/CERN/TIGRE/assets/63751488/4fe0876e-3f7d-4aeb-aad1-ae26527cf265)

Is there a “best” way to go about debugging these types of issues? Should I even expect that getting things “close enough” should get even a bad reconstruction? Thanks!

**Comment by AnderBiguri** [comment]
Reconstructions are extremely sensitive to geometric inaccuracies, but generally a DSD/DSO mismatch is the one that is less important. 

How to debug? ah, I don't know. I have build enough heuristic from years at looking at images to start slowly thinking about what things could be, but I have not found a fixed way. 

Few things that my brain things when seeing your images:

1- They've gone super bad. Negative values in the image are bad!
2- Above is quite possibly enhanced/caused by all those zeroes in your projection data. Zero is a valid number. FDK does not suffer from this, because it only takes the high frequency values of data, but any iterative algorithm will enforce detector values. If they see zeros, it can only be caused by negative values in the image (to compensate the other views where the detector has a value), and a negative value means a source of X-rays, which a patient should not be. Basically you need to figure out a way to either inpaint or crop all that area if you want to use iteratives. 
3- maybe your angles are negative sign. You get concave artifacts with edges pointing outwards, this often suggest some sign change (or projection flip) required. 
4- Your projections are not log transformed. TIGRE (and any recon) requires X-ray absorbtion (i.e. after Beer-law) to reconstruct, not X-ray transmission (i.e. raw measurement, light amount), otherwise its not a linear problem. The real data demo has some code to do so. You want "air" to be black, "bone" to be white. 


I don't fully understand your geometry diagram, so not sure what else to suggest, but if you explain how it differs from a standard circular trajectory, I may have suggestions on changing the TIGRE geo.

**Reply by ajensen1234** [reply]
1 - Interesting. Thanks for that gut-check!
2 - Would this be fixed by inverting the images (so bones are white), or padding those areas to be white instead of black? It seems like with cropping, those regions would still need to be padded.
3 - I am calculating all positive angles (see the gif of the sequence of projections)
4 - Which demo code performs this? I couldn't find anything.

I have done the following:

```python
projections = (4095 - projections) # invert the images
projections = np.log(projections).astype(np.float32)
```
This gives the following projection images:
![hip_moving](https://github.com/CERN/TIGRE/assets/63751488/d79d6914-a40f-4f99-91d9-52ca1cd52687)


Regarding the geometry, I will likely need to draw something out (I will attach later), but in words (referencing both the C-ARM figure and the matplotlib diagram)

- The center of rotation is determined by the C-shaped component on the OEC machine (measured with "Marker 74" in the figure).
- The source ("Focus trajectory" in the figure) is slightly offset from this center of rotation (effectively, the line from source to detector doesn't pass through the center of rotation of the machine, it merely traces a chord through the circle). Looking at the figure, this is demonstrated by the "focus beam start" and "focus beam end" not intersecting with the "center of rotation". Would this also change the detector location that needs to be specified?
- The location of the image/voxels/volume is also not at the center of rotation, but marked with a "*" (phantom location).

My thought process:
- The offset source/detector from center of rotation can be captured with `geo.COR`, because this emulates the "object" spinning, rather than the source at detector spinning
- The phantom offset would be handled with a consistent `geo.offOrigin` of the phantom location relative to the location of the center of rotation. (is this altered by having both `geo.offOrigin` and `geo.COR`?)

Remaining questions:
- Is the origin of the geometry ($\vec{v}_{orig}$ in the dissertation) pointing to voxel (0,0,0), or is it pointing to the center of the volume, which would span -0.5 to +0.5 * `geo.sVoxel` in all directions?


Thanks for the quick response! I know this is quite a few questions, and I appreciate your taking the time to mull it over and help out!

**Reply by AnderBiguri** [reply]
That log transform is wrong. Its the Beer-law that you need to apply:
```
proj=-np.log(proj/(np.max(proj+1)); # Beer-Lanbert law
```
You can replace `(np.max(proj+1)` by `2**12` in your case. 

For the angles, a rotation in a gif doesn't show you the direction in CT, could be either sign. I meant you may need to do `angles= - angles`. 

For the geometry, it seems you are in the right path, all that you say sound reasonable. I'd say ignore `geo.offOrigin` for now, it only defines which area of the world you reconstruct, worst case scenario your image is not in the center of the recosntructed volume. Not touching it removes a parameter from debugging, leave it for the last. 

The origin of the geometry is indeed the geometric center of the cube to reconstruct, generally not falling inside a pixel center (because sizes are almsot always even, thus the center will be in the edges between pixels).

**Reply by ajensen1234** [reply]
Hi Anders,

After applying the log transform, it seems to give reasonable results for single images (`plt.imshow(np.flipud(all_images[35]), cmap="grey")`):
![log_transform_hip](https://github.com/CERN/TIGRE/assets/63751488/b0bc4ade-3940-45e5-bf86-0c15b101fede)

But when I pass this same series of images to the `plotproj` function, it seems to be getting oversaturated:
`tigre.plotproj(all_images, angles.reshape(all_images.shape[0],1), savegif = "hip_moving.gif")`

![hip_moving](https://github.com/CERN/TIGRE/assets/63751488/70a76d5d-176d-460e-9476-5e1f6fab1315)

And the reconstruction seems to be a bit blown out as well:
```python
imgCGLS = algs.cgls(all_images, geo, angles, 10)

tigre.plotimg(imgCGLS,dim="Z",savegif="cgls.gif")
```
![cgls](https://github.com/CERN/TIGRE/assets/63751488/ea1c2d24-8793-4be8-9811-86a198be86a4)

Additionally, I've added a `+1` to the log transform to avoid any divide by zero errors:
`-np.log(image_data+1/4096).astype(np.float32)`

I've also replaced the blank values in the images with the average value of all the images in the dataset.


In some of the slices in the reconstruction, I can't tell if I'm hallucinating a few "shadows" of the pelvis image present, as well as a circle inscribed in a square (the exact same shape of the image output with the circle of the images and the 0s around it). I'm not sure if this is "normal" because of the images, and possibly improper geometry definitions, or if I've been staring at these images for so long that I'm trying to convince myself that something is there.

p.s. Can I buy you a beer/coffee? You've taken time out of your day to answer some of these questions, it's the least I can do.

Thanks!

**Reply by ajensen1234** [reply]
To minimize the complexity, we think that we are going to take some images at a few fixed points around the imaging trajectory (i.e. 7 total images at 0,15,30,45,60,75,90) just to simplify all possible variables. I'm hoping that this can also help guide the path forward for reconciling geometry information.

I will be sure to post any updates for any future researchers struggling with a similar challenge.

**Reply by ajensen1234** [reply]
Ahhh, okay, I think I was filtering images at the wrong spot.
Here's what I currently have, and it has achieved the best results so far.
```python
all_images = ds.pixel_array
all_images = np.where(all_images == 0, np.mean(all_images), all_images)
all_images = log_transform(all_images)

def log_transform(image_data):
    return -np.log((image_data + 1)/4096).astype(np.float32)
```
This gives the following transformed images:
![hip_moving](https://github.com/CERN/TIGRE/assets/63751488/aa7dfa46-e2ec-4555-b2fc-ce20788a180d)

These appear much more in line with what the demo images look like.

The reconstructions are still a bit strange, but I definitely see the pelvis in the images, so it feels a lot closer now:
![cgls-x](https://github.com/CERN/TIGRE/assets/63751488/10ef3f99-fe0b-4c19-b066-d9775c5989a9)
![cgls-y](https://github.com/CERN/TIGRE/assets/63751488/f96a791e-c4ff-4157-af4e-dc6a8da4636d)
![cgls-z](https://github.com/CERN/TIGRE/assets/63751488/9b8d274c-1588-4482-ad6d-5d2906dc5e75)

One thing that I still notice is that some reconstruction voxels are still going negative, which you mentioned was not a good thing — any thoughts on why that might be happening? I think I still have a bit of error in the geometry definition, could that be a source of this? There also appear to be a few border artifacts at the edges (I'm hoping that masking would have mitigated this).

**Reply by AnderBiguri** [reply]
Fantastic, good progress!
I'd say: make sure you recon with FDK first, before using the other algos. The negative values are almost surely because of that circular crop you have in the projection. As I mentioned before, a value of zero is as important as any other value for iterative recon, it means that there was no X-rays measured there, and the algos will enforce this. However, its not a real measurement, so it will  conflict with the other data (e.g. +90 deg) and the algorithms will try to add negative values in the pixels (meaning they are source of X-rays) to compensate. This is not an easy issue to fix, given your data... I'd say still try to recon with FDK first, as FDK won't have this issue, so any issue in FDK is because you are doing something wrong in the geometry definition.

**Reply by ajensen1234** [reply]
The FDK really helps a lot getting it pretty close. I think the only remaining steps are whittling down the geometry issues. At that, I had a few questions about how the geometry works:

1. What is the TIGRE x-axis? Currently, I have the x-axis as the vector from the center of rotation to the first location from the source, the z axis defined as positive rotations for the source, and the y-axis as $z \times x$.
2. I'm a little confused about how the center of rotation offset plays into the reconstruction. In our imaging setup, the vector from source to detector does not pass through the center of rotation of the machine. This is relatively the same thing as the object spinning ("Finally the centre of rotation (COR) offset has also been implemented, a common offset in CT machines where the sample rotates"), but with a partial arc scan, wouldn't it be important to define the COR with more than a single number (relative to the center of rotation)? So I guess the question here would be: how is the center of rotation defined in TIGRE? 
3. Is the TIGRE origin the center of rotation? Presently, my DSO is based on the radius of the rotation, which isn't the same distance as the source to the center of rotation.

Looking at some reconstructions in the z direction, I am starting to get a bit of intuition for the geometry, but it isn't quite perfect yet.
Thanks again!!!

**Reply by ajensen1234** [reply]
An additional thought on the center of rotation.

Imagine that your object is indeed the center of rotation. In order to specify the location of the object, we have $\vec{v}_{orig}$, but this could be either in the positive or negative y direction relative to the DSD vector. However, the COR value (based on my understanding) does not seem to discriminate between an offset in the positive or negative y direction.

Here's a top down view:
![image](https://github.com/CERN/TIGRE/assets/63751488/a32e48fd-c256-4912-97d7-bfbdf04d072a)

I'm going to be see if I can figure out why this interpretation might be wrong by pushing things out to extreme values, but this initial gut check is a little bit mystefying.

**Reply by ajensen1234** [reply]
Also, just to give as much context as possible, I've attempted to recreate the imaging setup using the different demo programs.

I have been editing `d14_Offsets.py`

```python
# Offsets
## ####################################################################################
# Lets try simple offset: The detector gets completely displaced
geo.offOrigin = np.array([-17.73541387, -154.73937889,   60.46204528])  # Offset of image from origin   (mm)
## ####################################################################################
# Auxiliary
geo.accuracy = 0.5
# Accuracy of FWD proj          (vx/sample)
geo.COR = -121
## Load data and generate projections
# see previous demo for explanation
angles = np.load("Python/demos/angles.npy")
# Load thorax phatom data
head = sample_loader.load_head_phantom(geo.nVoxel)
projections = tigre.Ax(head, geo, angles)


#%% lets see it
tigre.plotproj(projections, angles.reshape(-1,1))
res = algs.fdk(projections, geo, angles)
tigre.plotimg(res, dim="x", savegif="head-x.gif")
tigre.plotimg(res, dim="y", savegif="head-y.gif")
tigre.plotimg(res, dim="z", savegif="head-z.gif")
```

Here are the projections:
![moving-head](https://github.com/CERN/TIGRE/assets/63751488/a9bd5315-65b8-476c-9000-1083010035b2)

I need to dig to make sure that these are lining up with what we are actually doing here in the lab (might be a sign error somewhere), but they are still quite reasonable.

Here are the reconstructions:
![head-x](https://github.com/CERN/TIGRE/assets/63751488/185b6a4b-c92d-4792-a552-bfd68ffb1c0f)
![head-y](https://github.com/CERN/TIGRE/assets/63751488/625fdfc3-0cc8-4bdd-9363-cd322d17d375)
![head-z](https://github.com/CERN/TIGRE/assets/63751488/daf5bbf6-80b4-4dc8-a24b-4fedefea35cb)

I am still noticing a few aspects of the same negative values in the reconstructions, as well as some of the different "whispy" artifacts, which I wouldn't expect because we are now reconstruction from exactly the same geometry that we are projecting. Some of the sources that I've been reading say that this might be due to inherent limitations in the FDK algorithm. In our data, we only have about a 70 degree arc of images. I do think we are narrowing down on the geometry of the situation though, which might mean that it could be time to start looking into some of the iterative algorithms again.

Plugging in a larger arc of angles (where the head actually goes out of the frame a little bit), we see the following, which seems to me to indicate that some of the whispy artifacts are coming from an under-determined system into the FDK algorithm.

```python
angles = np.linspace(0, 2 * np.pi, 100)
```
![moving-head](https://github.com/CERN/TIGRE/assets/63751488/48cf4597-b5c9-4099-94e2-66aa075227f3)
![head-x](https://github.com/CERN/TIGRE/assets/63751488/0f66fc96-cb64-4eb2-b979-3116806cab10)
![head-y](https://github.com/CERN/TIGRE/assets/63751488/74420efa-44d9-4e9f-9eef-0d16cfd0b70d)
![head-z](https://github.com/CERN/TIGRE/assets/63751488/39773f5f-230e-42a6-b038-81d99f1d7012)

These appear much closer to "normal", with the regions of weirdness coming from the fact that the head goes out of view a little bit.


Also, based on the reference frames that I calculated, it was inputting a negative value for the center of rotation offsets to get the head projections to line up.

Thanks again for all the help!!

**Reply by AnderBiguri** [reply]
Apologies, this is a lot of things that you posted, and I am away, so let me give you the description of offOrigin and COR, and if this doesn't help you can ask more specific questions.   

Firstly, remember that these are in mm, not voxels. 

A normal CT scanner has its axis of rotation exactly in the cross of the source to **geometric center** of detector line, at distance DSO. At angle=0 this distance is also the x-axis. The displacement of the axis w.r.t. the source to center of detector line is COR. In a machine that rotates, the COR =0 by almost definition, as the machine rotates around a certain point, and that is it. However, if instead the machine is static, and there is a rotating stage with a sample on top (like in a microCT machine, common for scientific/industrial CT) then it is possible that the rotation of the sample does not happen in this line, due to mechanical inaccuracies of the location of the rotating sample (generally caused by the weight of the sample deforming the stage). This offset can of course happen in X and Y directions, but the X direction one is not necessary to correct, as that is the same as having a little bit different DSO, which is fine. However, the one in Y direction is crucial to correct. This tends to be a very small number anyway, something of the order of one or two pixels (in mm) maximum. 

Therefore, if your machine rotates, but not the sample, and this rotation axis is expected to be static (unless the machine is deforming and moving, this should be true) you do not need to set geo.COR at all. 

geo.offOrigin is easier to understand. Basically helps you select an area of the world to reconstruct that is not necesarily centered in the axis of rotation. This is a user choice and does not have a relationship with the scanning geometry, unless the sample/scanner is moving while you scan, like in a helical scan. 

Hope this helps! Let me know what you dont' understand

**Reply by ajensen1234** [reply]
Hi Ander, 
Thanks for the reply. Everything that you said makes sense. I think the geometry of the OEC9800 might be what is causing some of the issues that we might be having. 

Here's a schematic that demonstrates what is going on in our scans as the machine is rotating. 
![Projection Geometry - 5-7-2024](https://github.com/CERN/TIGRE/assets/63751488/4c96ea51-2468-456e-a4af-7a57cff30e8e)
 
Basically, we have a roughly 121mm (y direction) offset between the center of rotation of the machine, and the beam line that going from source to detector. To be sure, this is NOT a normal CT scanner setup, but rather an attempt at some interesting reconstructions.

We are in the process of removing the circular crop that is coming out of the images so that we can use some of the iterative algorithms, as they seem to work a little better with limited arcs of data (we don't get a full 180/360 degree scan, just about 75 degrees or so).

Does this seem plausible in your estimation?

**Reply by AnderBiguri** [reply]
Interesting! 

It seems indeed that COR is what you want to be touching indeed. Is 121mm the right number? That is a significant portion of the DSD, right? 

If your machine at rotating 360 (I know it cant), would produce a set of blue lines (in the right figure source-detector) that make a circle, and never cross the red origin, indeed I think COR is the one you need to be thinking about. I always thought about it with a static machine and rotating sample, which makes it easier to visaulize, but I think its the same thing given a change of reference frame.

**Reply by ajensen1234** [reply]
To obtain this offset value, we took a few measurements. The first was the actual center of rotation of the machine (we calculated multiple circles from taking subsets of 3 points from the trajectory). Second, we determined the DSD vector from the source to the detector. 121 was calculated as the minimum value between the COR and this DSD vector, or the length of the perpendicular vector from the DSD that intersects with the COR. This was validated at multiple positions along the C-arm trajectory.

The DSD is coming in at around 1200mm, so this offset is roughly 10% of the DSD. 
On your remark that the blue lines would sort of trace out an interior circle, rather than cross the red X, you are absolutely correct. This was causing quite a bit of pain when I was trying to figure out the geometry without assuming that there was a COR offset in our model.

I've also done quite a few sketches to myself to prove that this is indeed the same relative reference frame as an object spinning  121mm away, albeit in the opposite direction. 

Additionally, we did set the COR to be negative because you can imagine that the direction of the offset would have an effect on the images that are getting generated. 

I think I've had all my questions squared away. I will let you know how the reconstructions work once we are dealing with images that are better suited for the iterative algorithms.

**Reply by AnderBiguri** [reply]
Fantastic! Do also feel free to drop me an email if that helps :)

**Reply by ajensen1234** [reply]
I sure will. I'll keep this discussion open to make sure that all new forms of information stay in the same place. Thanks again for all the help!

**Reply by ajensen1234** [reply]
So, we were finally able to get images that do not have the circular crop, so the iterative methods seem to giving reasonable answers that are debuggable. 

Saving the volume from the `plotimg` function tends to compress a lot of the details, but here are the following image sequences and reconstructions.

![hip_moving](https://github.com/CERN/TIGRE/assets/63751488/c0f5f4f9-6f96-48cc-8f51-8b23104fad3b)
![fdk-x](https://github.com/CERN/TIGRE/assets/63751488/afda0235-ddce-4f2c-b484-882915e84e97)
![fdk-y](https://github.com/CERN/TIGRE/assets/63751488/2ad4506e-3aac-4d48-b372-3322f144c652)
![fdk-z](https://github.com/CERN/TIGRE/assets/63751488/1bd32923-6329-49e8-8775-96bf80488e26)

I wonder if you have any visual intuition on what might be going on here? I think the clearest case of geometric “weirdness” might be in the z direction of the reconstruction. At around frame 70, you'll notice some lead washers appear in from the left side of the image and move across to the right side of the image. Clearly, this means that the reconstructed box is not properly aligning with the actual geometry of the images as they were taken. 

I'm going to be taking a lot of measurements over the following days to make sure that I'm getting everything right regarding the relative positions, but I was hoping that you might have built up some intuition on some quick fixes to try.

**Reply by ajensen1234** [reply]
I've finally created a visualizer that takes as input a geometry/angle definition (at least for a planar rotation, not an arbitrary rotation)

Here is a short video showing this. 


https://github.com/CERN/TIGRE/assets/63751488/cf47abae-f0f6-401f-9725-dcb065a6d6ee

I'll be using this to try and debug any issues that I'm having with the geometry and projections showing up in weird places

**Reply by AnderBiguri** [reply]
For those volume visualization: I suggest to set the colorbar range manually, in particular do not let negative values appear. May help you see better, as such a limited angle scan will certainly produce a relatively bad image.

**Reply by ajensen1234** [reply]
On it producing a relatively bad image, I read the following paper:
[Gross et al_2010_GPU-Based Volume Reconstruction from Very Few Arbitrarily Aligned X-Ray Images.pdf](https://github.com/CERN/TIGRE/files/15339282/Gross.et.al_2010_GPU-Based.Volume.Reconstruction.from.Very.Few.Arbitrarily.Aligned.X-Ray.Images.pdf)
Which was claiming to get quite good reconstruction from as few as 5 images using conjugate gradient least squares with ridge and total variation regularization. Any intuition on that?

**Reply by AnderBiguri** [reply]
@ajensen1234  those are relatively bad recons, but they also are very specifically not limited angle (which makes the problem much harder) and with simulated noise in general. Still, they almost surely also control the visualization range, to be able to hilight the ct values where the information is ;)

**Reply by ajensen1234** [reply]
That makes sense. I'm a bit new to the world of these types of reconstructions, so I don't really have an intuition for what “good” and “bad” reconstructions are. It also makes sense that they would be able to have a bit of an easier time with it because they are defining the projection parameters in both the dataset creation and the reconstruction portions, so it's something of a closed system. 

I'll keep hacking away — thanks again!

**Reply by AnderBiguri** [reply]
@ajensen1234 indeed! One of the things I've definetly experienced is that you certainly will get less fancy looking results when using real data. Real data is quite more complex and prone to geometric and complex noise than simulated one, so always assume worse performance in it than the fine tuned, sometimes cherry picked, paper results :D

**Reply by ajensen1234** [reply]
Some of the reconstruction parameters still seem to be causing some streaking in the z slices. The visualizer helps, but we also think that there might be some mechanical "slop" in the machine that would be impossible to measure.

To keep you updated on this, we are planning to create a cube phantom with some exactly placed BBs on each face as fiducial markers. From there, we have some folks in the lab that have quite a bit of experience with calibrating intrinsic robot parameters for kinematics tracking. We are going to use a similar toolkit to try and determine the reconstruction parameters that best recreate our known geometry.

I guess a question I would have, if you look at the z slices  (starting at slice 125) of the reconstruction (below), do you have any intuition on which parameter would be causing the objects to move across the screen like they are? I can see the viewing frustrum intersecting with the volume based on the geometry, but I'm unsure what might be causing the images to streak across like that. 


https://github.com/CERN/TIGRE/assets/63751488/ffe879a8-b1bb-4d9f-8f78-f415eb2eb554


https://github.com/CERN/TIGRE/assets/63751488/8412631d-d458-4ba6-ae33-f951cda0ecaf


Thanks again, and I'll keep you posted on any of the updates that we have using some of the robot optimization techniques.

**Reply by AnderBiguri** [reply]
@ajensen1234 first, I would stilll suggest that you do limit the colorbar axis to only positive values, and not allow negatives! It will make viualization much better. Also from above, often CT creates 1 or 2 pixels with very high value, so contrast is removed from images. 

The effect you are describing is basically the difference between an area that there is information to add in the volume, and area where is not. that "traversal" is only due to the cone shape, i.e. everything of uniform value outside is never covered by the blue cone in your second gif.

**Reply by ajensen1234** [reply]
The recon plots are coming out of the tigre library `plotimg`, I will edit the values that go below zero to zero.
```python
imgCGLS = algs.cgls(all_images, geo, angles, 10)
imgCGLS = np.where(imgCGLS < np.mean(imgCGLS)*10, imgCGLS,np.mean(imgCGLS)*10 )
imgCGLS = np.where(imgCGLS>0, imgCGLS, 0)
```

https://github.com/CERN/TIGRE/assets/63751488/31e65c72-454d-4dd7-a31c-5f3550471de2


https://github.com/CERN/TIGRE/assets/63751488/72ba3326-3c88-426f-89f7-298f5d5ec192


https://github.com/CERN/TIGRE/assets/63751488/9761544b-0e80-40d1-a8d1-99128038ae42



Regarding the traversal and the cone shape, I understand that the whispy shapes are coming from the cone sweep. Specifically though, if you look at frame 142, a series of 4 washers appear on the left side of the reconstruction, and they leave the image at slice 178 on the right side. 

In our setup, the washers are rigidly fixed to the phantom, and so we would expect to see this z-slicing show them in the same x/y position, rather than moving across the screen. It's due to these recon errors that we are assuming that some optimization needs to be done on the geometric properties.

**Reply by AnderBiguri** [reply]
It looks like your geo.offOrigin is wrong, as you are doing the recon only in a quarter of the volume.
The explanation for your washers is the same, really. If you only see them in 1 projection, then at backprojecting they will be smeared out all along the ray path. Unless you see them in all projections over a circle, you will alwasy have this smearing to some degree. That is what the "stripe artifacts" of sparse angle/limited angle tomograhy is anyway. The backprojection will put the info of the X-ray all along the path equally. It is _not seing it_ in the other views that removes this info, and if you have a limited angle tomographic problem, it is a lot of not _not seing_ that you are doing, if that makes any sense.

**Reply by ajensen1234** [reply]
Hi Ander! Just to keep you updated on what we've been working on!

We 3D printed a calibration cube with known lead bead locations, and took a series of images during a fix machine arc. Then, using TIGRE variable's we construct a projection function and compare the projected location to the marker locations in each image. Using some pretty standard optimization algorithms, we should be able to fine-tune the TIGRE parameters to fit our machine exactly!

Still making sure that all the math is doing what we expect it to do, but we seem to be able to test all our assumptions about the geometry this way, and should end up with a much nicer reconstruction!

I'll continue posting here as we achieve any wins or milestones!

**Reply by ajensen1234** [reply]
Hi Ander. Just wanted to pop in with some updates, and some questions.

We've created a calibration cube and used it to quite accurately determine the projection parameters of our camera, as well as the relative motion between our cameras.

One thing that we are having trouble with is reconciling the TIGRE geometric definitions (COR, detector offset, etc) with a standard camera projection matrix with intrinsic and extrinsic coordinates defined. If you could  point me to the location in the code where you are actually performing those projections, and moving the camera around, that would be immensely helpful to be able to ensure that we are representing things in the correct way.

Let me try and write out what I've done to try and homogenize the two different representations.

EDIT: I'm unsure how to get mathematical symbols rendering properly in GH's latex syntax, I hope the explanations are clear.

First, as the camera is moving through it's arc, TIGRE defines this with a COR, DSO, and angle. In standard extrinsic coordinates, this can be defined as a single transformation matrix.

```math
T^A_B = \begin{bmatrix} 
  & R_{3 \times 3} & & T_{3 \times 1} \\ 
  0& 0 & 0 & 1 
\end{bmatrix}
```
The equivalent transform in the TIGRE system is the equivalent of translating the camera (DSO,COR,0)(DSO,COR,0), rotating by θθ, then translating $(−DSO,−COR,0)(−DSO,−COR,0)$.

Which would give the equivalent transformation of: $T_ × R × −T$, where $T$ is a positive translation, $R$ is a rotation about the TIGRE z axis, and $-T$ a negative translation. Am I understanding this correctly?

And so, to determine the relative offset, we were going to solve this as a system of equations and determine the exact values for the rotation and the DSO/COR.

**Reply by ajensen1234** [reply]
Ah, I think I've found it.

It looks like it's in `ray_interpolated_projection.cu` in line 632 under `computeDeltas`, which populates the locations of the source and the detector for a given series of rotations. This should be super helpful in getting this exactly correlated.

If needed, I also might see what I can do making a branch that has a `computeDeltas` that takes just the extrinsic parameters of the camera as a transformation matrix to see if that makes it any easier.

Please let me know if I seem to have missed anything!

**Reply by AnderBiguri** [reply]
Hi @ajensen1234 . You are correct, but caution! `computeDeltas` does transfor everything to a strange "projection geometry" in where the image never rotates and each pixel is of size [1,1,1], and the corner pixel sits in the origin. Its not an intuitive transform at all. Also TIGRE has more degrees of freedom than required, as it makes it easier to feed the info in a "CT" way. That unfortunately means that indeed getting from camera parameters to TIGRE geometry is non trivial (and for some cases-non unique). 

That said, if you do end up getting a function that can do it, it would be a great addition to TIGRE.

**Reply by ajensen1234** [reply]
Hi Ander, 
Thanks for the reply!
I think I was able to figure out some of the dimensional transforms and shifts going on (I'm guessing having the dimensions end up "unitless" makes getting things into CUDA kernels a lot easier!)

One thing I was a little bit confused about was defining the location of the source after a rotation.
In the order of the source getting interacted with in the fuction:


1. Create source object, and move it along the x-axis to DSO (lines 633-636)
```cpp
    Point3D S;
    S.x=geo.DSO[i];
    S.y=0;
    S.z=0;
```

2. Rotate the object by $\theta$ (line 694). For now, we'll assume just one rotational degree of freedom, so it ends up just being a z rotation.
```cpp
    eulerZYZ(geo,&S);
```

3. Negative source offset (same relative transform as positive image offset) (line 702)
```cpp
S.x=S.x-geo.offOrigX[i]; 
S.y=S.y-geo.offOrigY[i];  
S.z=S.z-geo.offOrigZ[i];
```

4. Adjusting pixel locations to get the center of the image as the origin and scaling units (708, 714)
```cpp
S.x       =   S.x+geo.sVoxelX/2-geo.dVoxelX/2;           
S.y       =   S.y+geo.sVoxelY/2-geo.dVoxelY/2;               
S.z       =   S.z  +geo.sVoxelZ/2-geo.dVoxelZ/2;
S.x       =    S.x/geo.dVoxelX;           
S.y       =    S.y/geo.dVoxelY;            
S.z       =    S.z/geo.dVoxelZ;

```

5. Adjusting for center of rotation (lines 720-726). You do this by rotating a y-vector about the same rotation $\theta$
```cpp
    float CORx, CORy;
    CORx=-geo.COR[i]*sin(geo.alpha)/geo.dVoxelX;
    CORy= geo.COR[i]*cos(geo.alpha)/geo.dVoxelY;
    Pfinal.x+=CORx;   Pfinal.y+=CORy;
    Pfinalu0.x+=CORx;   Pfinalu0.y+=CORy;
    Pfinalv0.x+=CORx;   Pfinalv0.y+=CORy;
    S.x+=CORx; S.y+=CORy;
```

So, if we remove the dimensional scaling, and the relative transforms that are occuring to account for the image offset, the transforms that soley describe the location of the detector is broken down into the $\theta$ rotation and the center of rotation offset.

However, I think there's either (1) and error in the way that the final location of the source is being determined, or (2) I have a misunderstanding of what the center of rotation offset means.

Regarding #2, my understanding of the center of rotation offset is that there is a point that does not lay on the `TIGRE` x-axis about which the source-detector unit is spinning. 

Here's a diagram:
![image](https://github.com/CERN/TIGRE/assets/63751488/3e016e28-c1e0-4bee-b8d9-f34ff3dd6a72)

For this understanding of the center of rotation, the equivalent transformation is: 

1. Move the source by `COR`.
2. Apply the rotation
3. Move the source by `-COR` along the rotated axis.

Here's a diagram:
![image](https://github.com/CERN/TIGRE/assets/63751488/8839f693-b46d-4448-83ae-e694352a144d)


And so, if my understanding of both a rotation about COR is correct, and I've read the `computeDeltas` correctly, then I think there's an issue. Mainly, that the application of this rotation is just:
1. Rotate $(DSO, 0, 0)$ about $\theta_y$ (line 694)
2. Apply rotated COR vector (lines 720-726)

Here's a diagram:
![image](https://github.com/CERN/TIGRE/assets/63751488/168d85d3-ef68-45a3-ab1b-1f7643b14b25)

 It's distinctly possible that I'm missing something here, and perhaps some of this is getting handled in the dimensional adjustments that I didn't quite pick up. I do think, if your COR was sufficiently small, then these effects would probably not be noticable. But, unfortunately for us, we had a large COR, so my diagrams just assumed that it was big. The math would remain the same, though.

Please do let me know what you think; I'm sorry the message is so long, but I wanted to be thorough.

**Reply by AnderBiguri** [reply]
I'll try to check all this logic when I have a little bit of time. I tend to just lose track of what the exact math is here, its clearly non-trivial and not super easy to follow. However, its extremely unlikely there is a mistake, as I have verified my operators with many many other recosntruction tools over the years, and I can get the same results given the same inputs.

**Reply by ajensen1234** [reply]
That sounds good. Thanks ahead of time for taking the time to look it over!

**Reply by ajensen1234** [reply]
Hi Ander, 
I just found the ASTRA toolbox (https://github.com/astra-toolbox/astra-toolbox), which seems to support some arbitrary geometries, at the cost of less flexibility in the algorithm construction. Having done so, I found the following issue #515 and commit https://github.com/CERN/TIGRE/commit/66cfc437ff6a9286d6e2523112e97305d1e66c09.

I think this might be exactly the solution that I was looking for, as it was a bit more straightforward to directly calculate the transformation matrices of our system. 

I will let you know if this works. If it does, I can close this discussion, and hopefully all the dialogue will be useful for anyone else in the future trying to do fun things with CT reconstruction!

**Reply by AnderBiguri** [reply]
Hi @ajensen1234 . ASTRA==TIGRE for flexibility of geometries, the only difference really is how they are defined :) 

I had completely forgotten about that PR. Indeed, that seems to be the thing you need. Do let me know if it works, I never tested it myself!
